crates/goose/src/agents/agent.rs

```diff
@@ -1587,4 +1597,142 @@ mod tests {
 
         Ok(())
     }
+
+    enum NamingBehavior {
+        Success,
+        Error,
+        Panic,
+    }
+
+    #[derive(Clone)]
+    struct MockNamingProvider {
+        model_config: ModelConfig,
+        behavior: Arc<std::sync::Mutex<NamingBehavior>>,
+        captured_session_id: Arc<std::sync::Mutex<Option<String>>>,
+    }
+
+    impl MockNamingProvider {
+        fn new(behavior: NamingBehavior) -> Self {
+            Self {
+                model_config: ModelConfig::new_or_fail("test-model"),
+                behavior: Arc::new(std::sync::Mutex::new(behavior)),
+                captured_session_id: Arc::new(std::sync::Mutex::new(None)),
+            }
+        }
+
+        fn get_captured_session_id(&self) -> Option<String> {
+            self.captured_session_id.lock().unwrap().clone()
+        }
+    }
+
+    #[async_trait]
+    impl Provider for MockNamingProvider {
+        fn metadata() -> crate::providers::base::ProviderMetadata {
+            crate::providers::base::ProviderMetadata::empty()
+        }
+
+        fn get_name(&self) -> &str {
+            "mock"
+        }
+
+        async fn complete_with_model(
+            &self,
+            _model_config: &ModelConfig,
+            _system: &str,
+            _messages: &[Message],
+            _tools: &[Tool],
+        ) -> anyhow::Result<(Message, ProviderUsage), ProviderError> {
+            Ok((
+                Message::assistant().with_text("Response"),
+                ProviderUsage::new("mock".to_string(), Default::default()),
+            ))
+        }
+
+        async fn generate_session_name(
+            &self,
+            _messages: &Conversation,
+        ) -> Result<String, ProviderError> {
+            *self.captured_session_id.lock().unwrap() =
+                crate::session_context::current_session_id();
+
+            let behavior = self.behavior.lock().unwrap();
+            match *behavior {
+                NamingBehavior::Success => Ok("Generated Name".to_string()),
+                NamingBehavior::Error => {
+                    Err(ProviderError::RequestFailed("naming failed".to_string()))
+                }
+                NamingBehavior::Panic => panic!("naming panicked"),
+            }
+        }
+
+        fn get_model_config(&self) -> ModelConfig {
+            self.model_config.clone()
+        }
+    }
+
+    // Verifies session ID is visible in generate_session_name when user hasn't provided a name.
+    // When user provides name, maybe_update_name early-returns and session ID isn't captured.
+    #[test_case(NamingBehavior::Success, None, "Generated Name", true)]
+    #[test_case(NamingBehavior::Error, None, "initial", true)]
```

Comment:
error and panic cases here are just to make sure our ignoring the same doesn't accidentally crash the critical code. we didn't have a problem so far, but it is easy to create one, so these cases help IMHO